modules/imgproc/src/smooth.cpp

```diff
@@ -3265,58 +3276,73 @@ class BilateralFilter_8u_Invoker :
-
-                            _w = _mm_hadd_ps(_w, _g);
-                            _mm_store_ps(bufSum, _w);
-
-                            wsum  += bufSum[0];
-                            sum_b += bufSum[1];
-                            sum_g += bufSum[2];
-                            sum_r += bufSum[3];
-                         }
+                            const uchar* const sptr_k0  = sptr + j + space_ofs[k];
+                            const uchar* const sptr_k1  = sptr + j + space_ofs[k+1];
+                            const uchar* const sptr_k2  = sptr + j + space_ofs[k+2];
+                            const uchar* const sptr_k3  = sptr + j + space_ofs[k+3];
+
+                            v_float32x4 __b = v_cvt_f32(v_reinterpret_as_s32(v_load_expand_q(sptr_k0)));
+                            v_float32x4 __g = v_cvt_f32(v_reinterpret_as_s32(v_load_expand_q(sptr_k1)));
+                            v_float32x4 __r = v_cvt_f32(v_reinterpret_as_s32(v_load_expand_q(sptr_k2)));
+                            v_float32x4 __z = v_cvt_f32(v_reinterpret_as_s32(v_load_expand_q(sptr_k3)));
+                            v_float32x4 _b, _g, _r, _z;
+
+                            v_transpose4x4(__b, __g, __r, __z, _b, _g, _r, _z);
+
+                            v_float32x4 bt = v_abs(_b -_b0);
+                            v_float32x4 gt = v_abs(_g -_g0);
+                            v_float32x4 rt = v_abs(_r -_r0);
+
+                            bt = rt + bt + gt;
+                            v_store(buf, v_round(bt));
+
+                            v_float32x4 _w  = v_float32x4(color_weight[buf[0]],color_weight[buf[1]],
+                                                    color_weight[buf[2]],color_weight[buf[3]]);
+                            v_float32x4 _sw = v_load(space_weight+k);
+
+                            _w *= _sw;
+                            _b *=  _w;
+                            _g *=  _w;
+                            _r *=  _w;
+
+                            vsumw += _w;
+                            vsumb += _b;
+                            vsumg += _g;
+                            vsumr += _r;
+                        }
+#if CV_SSE3
+                        // use horizontal add
+                        float *bufFloat = (float*)buf;
+                        vsumw.val = _mm_hadd_ps(vsumw.val, vsumb.val);
+                        vsumg.val = _mm_hadd_ps(vsumg.val, vsumr.val);
+                        vsumw.val = _mm_hadd_ps(vsumw.val, vsumg.val);
+                        _mm_store_ps((float*)bufFloat, vsumw.val);
+
+                        wsum  += bufFloat[0];
+                        sum_b += bufFloat[1];
+                        sum_g += bufFloat[2];
+                        sum_r += bufFloat[3];
+#else
+                        wsum += v_reduce_sum(vsumw);
+                        sum_b += v_reduce_sum(vsumb);
+                        sum_g += v_reduce_sum(vsumg);
+                        sum_r += v_reduce_sum(vsumr);
```

Comment:
There is `v_reduce_sum4` with SSE3 replacement.

Could we store (or calculate) "wsum, sum_b, sum_g, sum_r" as `v_float32x4` ?

Reply:
```v_reduce_sum4``` !!!  I dreamed about this implementation and it exists !! wow !!

Reply:
You can also use `v_reduce_sum4` in that place (and, in fact, in other blocks which use `_mm_hadd_ps` in order to get rid of SSE3-dependant code).

Reply:
Indeed, your right.  I followed your comment and the code looks very nice with no more ```CV_SSE3```
Thanks !